File: TestDomain/RocketShip.cs
using FluentNHibernate.Data;
using lvl.Ontology;
using lvl.Ontology.Authorization;
using lvl.Ontology.Conventions;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace lvl.TestDomain
{
    [OwnedByUser(nameof(TargetPlanetId), Actions.Read | Actions.Update)]
    [Role("nasa")]
    public class RocketShip: Entity, IAggregateRoot
    {
        public DateTime ExpectedLaunchDate { get; set; }
        public decimal Cost { get; set; }
        public decimal MaximumWeight { get; set; }

        [Required]
        public string InventingUserId { get; set; }
        
        [ForeignKeyId(typeof(Planet))]
        public long TargetPlanetId { get; set; }

        public IEnumerable<Booster> Boosters { get; set; }
    }
}

File: front-end/demo-app/models/rocket-ship.ts
import { Entity, IAggregateRoot, Required } from '../../src';
import { Booster } from './booster';

export class RocketShip extends Entity implements IAggregateRoot {
    public expectedLaunchDate: Date;
    public cost: number;
    public maximumWeight: number;
    @Required() public inventingUserId: string;
    public targetPlanetId: number;
    public boosters: Booster[];

    constructor(options?: IRocketShipOptions) {
        options = options || {};
        super(options);
        this.expectedLaunchDate = options!.expectedLaunchDate;
        this.cost = options!.cost;
        this.maximumWeight = options!.maximumWeight;
        this.inventingUserId = options!.inventingUserId;
        this.targetPlanetId = options!.targetPlanetId;
        this.boosters = options.boosters ? options.boosters.map(p => new Booster(p)) : [];
    }
}

interface IRocketShipOptions {
    expectedLaunchDate?: Date;
    cost?: number;
    maximumWeight?: number;
    inventingUserId?: string;
    targetPlanetId?: number;
    boosters?: Booster[];
    id?: number;
}
